File: app/services/TmdbApiService.java
package services;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

import models.helper.TmdbMovie;
import play.Logger;
import play.Logger.ALogger;
import play.Play;
import play.libs.WS;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;

public class TmdbApiService {
	
	private static final String LANG_EN = "en";
	private static final String TYPE_JSON = "json";
	//private static final String TYPE_YAML = "yaml";
	//private static final String TYPE_XML = "xml";
	
	private static final String BASE = "http://api.themoviedb.org/2.1";
	private static final ALogger logger = Logger.of("application.tmdb");
	
	
	private final String apiKey;
	
	public TmdbApiService() {
		this.apiKey = Play.application().configuration().getString("tmdb.apikey");
	}
	
	public TmdbMovie findOrRead(final String title){
		return findOrRead(title, null);
	}

	public TmdbMovie findOrRead(final String title, final Integer year){
		if(title==null){
			throw new IllegalArgumentException("NULL title");
		}
		TmdbMovie tmdbMovie = null;//TmdbMovie.findByTitleAndYear(title, year);
		if(tmdbMovie == null){
			try{
				TmdbMovie[] result = search(title, year);
				if(result.length == 0 && year != null){
					// try a second search without year
					result = search(title);
				}
				if(result.length != 0){
					tmdbMovie = result[0];
					//tmdbMovie.save();
				}
			}catch(Exception ex){
				logger.error(ex.getMessage(), ex);
			}
		}
		return tmdbMovie;
	}
	
	
	private TmdbMovie[] search(final String title){
		return this.search(title, null);
	}
	
	private TmdbMovie[] search(final String title, final Integer year){
		String search = title;
		if(year != null){
			search = search + " " + year;
		}
		String encodedSearch = urlEncodeUtf8(search);
		// remove question mark ast this doesn't work
		encodedSearch = encodedSearch.replace("%3F", "");
		
		String url = BASE + "/Movie.search/" + LANG_EN + "/" + TYPE_JSON + "/" + apiKey + "/" + encodedSearch;
		logger.info(url);
		JsonElement element = new JsonParser().parse(WS.url(url)
				.setHeader("Accept","application/json").get().get().getBody());
		
		if(element.isJsonArray() && element.getAsJsonArray().get(0).isJsonPrimitive()){
			String message = element.getAsJsonArray().get(0).getAsString();
			logger.warn(String.format("%s returns %s", url, message));
			return new TmdbMovie[0];
		}
		
		Gson gson = new GsonBuilder().setPrettyPrinting().create();
		//System.out.println(gson.toJson(element));
		return gson.fromJson(element, TmdbMovie[].class);
	}
	
	String getToken(){
		String url = BASE+"/Auth.getToken/"+TYPE_JSON+"/" + apiKey;
		JsonElement element = new JsonParser().parse(WS.url(url)
			.setHeader("Accept","application/json").get().get().getBody());
		return element.getAsJsonObject().get("token").getAsString();
	}
	
	String auth(final String token){
		// redirect here
		String url = "http://www.themoviedb.org/auth/" + token;
		return url;
		
	}

	private String urlEncodeUtf8(final String string) {
		try {
			return URLEncoder.encode(string, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			throw new RuntimeException(e);
		}
	}
}

File: app/services/TmdbApiService.scala
package services

import play.api.libs.ws.WS
import play.api.Logger
import play.api.Play.current

import models.helper.TmdbMovieSearch
import models.helper.TmdbMovieSearchPager

import scala.concurrent.ExecutionContext.Implicits.global
import concurrent.Future

object TmdbApiService extends JacksonMapper{
	
	val BASE = "http://api.themoviedb.org/3"
	  
	private val logger = Logger("application.tmdb")

	private lazy val apiKey = PlayUtil.config("tmdb.apikey")

	def find(title:String, year:Option[Int] = None): Future[Option[TmdbMovieSearch]] = {
		if(title==null){
			throw new IllegalArgumentException("NULL title")
		}
		val result = search(title, year)
		// TODO fall back to title only search if none found?
		// create test to see if this is needed
		result.map(list => list.headOption)
	}
	
	
	def search(title:String, year:Option[Int] = None) = {
	  println(apiKey)
		val url = BASE + "/search/movie"
		
		val qs = List("api_key"-> apiKey, "query" -> title)
		val qs2 = year.map(y => qs.+:("year" -> y.toString)).getOrElse(qs)
		
		logger.info(url)
		val response = WS.url(url)
		  .withHeaders("Accept" -> "application/json")
		  .withQueryString(qs2:_*).get()
		  
		response.map{ r =>
		  logger.info(url + " " + r.status)
		  r.status match {
		    case 503 => 
		      logger.error(r.status + " " + r.statusText + " " + r.body)
		      List()
        case 401 =>
          logger.error(r.status + " " + r.statusText + " " + r.body)
          List()
		    case _ => 
		      val pager = deserialize[TmdbMovieSearchPager](r.body)
		      pager.results
		  }
	      
		}
	}
}
